src/snowflake/snowpark/_internal/code_generation.py

```diff
@@ -0,0 +1,393 @@
+
+    """
+    # order matters to classes_to_generate, when constructing source code, referenced classes need to be defined first.
+    classes_to_generate.insert(0, cls)
+    # get base classes first
+    inferred_clses = [v for v in reversed(cls.__bases__) if v.__qualname__ != "object"]
+
+    # __dict__ contains function, classmethod, classes attributes within a given class which
+    # needs to be further analyzed
+    clsdict = dict(cls.__dict__)
+    for v in clsdict.values():
+        if isinstance(v, type) and v.__module__ == func_module_name:
+            # v is a class defined in the same module as UDF func's, need to dynamically parse the class
+            inferred_clses.append(v)
+        elif inspect.isfunction(v) or isinstance(v, classmethod):
+            # v is a function/classmethod, get the references objects of the Function object
+            get_func_references(
+                v if not isinstance(v, classmethod) else v.__func__, ref_objects
+            )
+
+    # recursively handling classes that should be parsed dynamically
+    for more_cls in inferred_clses:
+        get_class_references(
+            more_cls, func_module_name, ref_objects, classes_to_generate
+        )
+
+
+def extract_func_global_refs(code: CodeType) -> dict:
+    # inspired by cloudpickle to recursively extract all the global references used by the target func's code object
+    co_names = code.co_names
+    out_names = {}
+    for instr in dis.get_instructions(code):
+        op = instr.opcode
+        if op in GLOBAL_OPS:
+            out_names[co_names[instr.arg]] = None
+
+    if code.co_consts:
+        for const in code.co_consts:
+            if isinstance(const, CodeType):
+                out_names.update(extract_func_global_refs(const))
+
+    return out_names
+
+
+def remove_function_udf_annotation(
+    udf_source_code: str, code_as_comment: bool = True
+) -> str:
+    if code_as_comment:
+        return udf_source_code
+    # remove the udf/pandas_udf annotation to avoid re-registration
+    res = re.search(r"@(udf|pandas_udf)\(", udf_source_code)
+    if res is None:
+        return udf_source_code
+    udf_anno_begin = res.start()
+    udf_anno_end = res.end()
+    cnt = 1
+
+    # find the pairing ')' of the leading 'udf('
+    while cnt != 0:
+        if udf_source_code[udf_anno_end] == "(":
+            cnt += 1
+        elif udf_source_code[udf_anno_end] == ")":
+            cnt -= 1
+        udf_anno_end += 1
+
+    return udf_source_code[:udf_anno_begin] + udf_source_code[udf_anno_end:]
+
+
+def check_func_type(func: Any) -> None:
+    if (
+        isinstance(func, classmethod)
+        or inspect.ismethod(func)
+        or not (isinstance(func, (FunctionType, BuiltinFunctionType)))
+    ):
+        logger.debug(f"Code generation for {type(func)} is not supported yet.")
+        raise TypeError(f"Code generation for {type(func)} is not supported yet.")
+
+
+def generate_source_code(
```

Comment:
For each step in this function, can we extract them to a function so it is easier to reason?

I think the major confusion for me is that I don't understand if each step has dependencies on previous steps, or they just happen to be in this sequence.

Reply:
yes, it's like a pipeline.

it's a piece of good suggestion to split them, I have updated the code to make them separate function calls.